Task/biagas/pr3456 update with develop

biagas opened the pull request:
This is to get @mathstuf's PR #3456 into develop.
I created a branch off #3456 to preserve original commit messages.
I merged develop into the new branch.
I fixed a segv that showed up on Windows.
I compiled on Windows and ran the test suite successfully.
I compiled on Linux and am currently running the test suite (one of the reasons I will mark this as a draft pull request.)

mathstuf committed:
utf-8: replace invalid non-utf8 sequences

mathstuf committed:
whitespace: add missing trailing newlines

Sven-Buijssen committed:
avtFVCOMReader: fix docstring typos

Commit:
64bit: use vtkIdType where necessary

Co-Author: Sebastien Jourdain <sebastien.jourdain@kitware.com>

Commit:
avtLAMMPSDumpFileFormat: fix the mesh type of unitCell

It is actually a point mesh, not an unstructured mesh.

jourdain committed:
vtkCEAucdReader: fix the data object set method usage

danlipsa committed:
visit_vtkOpenFOAMReader: prevent duplicate symbols within ParaView

utkarshayachit committed:
avtMiliFileFormat: add a missing include

danlipsa committed:
paraDIS: fix int64_t shadowing on Ubuntu

berkgeveci committed:
vtk: update to the latest timestep API

See r32871 for upstream's changes which mostly implemented this commit.
This hunk might still be relevant.

mathstuf committed:
vtk: update for the vtkMTimeType typedef

mathstuf committed:
warnings: fix overload-virtual warnings

mathstuf committed:
avtparaDIS_tecplotFileFormat: use the `std::` namespace

mathstuf committed:
vtk: update for vtkDataArrayTemplate changes

mathstuf committed:
DBOptionsAttributes: do not use as a reference-to-pointer

mathstuf committed:
paraview/RC_cpp_lib: fix C errors

Variables must be declared at the top of functions.

danlipsa committed:
paradis: fix a const correctness bug

cjh1 committed:
avtGenericDatabase: remove trivial producer

(Ben): This apparently does not work with the new VTK pipeline API.

Commit:
avtLAMMPSDumpFileFormat: properly read the IDs of the atoms

danlipsa committed:
Chombo, Enzo: fix HDF5 warning about the H5F_CLOSE_SEMI property

mathstuf committed:
databases: avoid symbol collisions for Atom structures

mathstuf committed:
databases/Vs: forward declare DBOptionsAttributes

Commit:
Adding support for reading in 10-noded tets for ANSYS .inp files

cquammen committed:
Enable reading of IEEE big-endian floating point arrays

The PFLOTRAN reader did not work correctly for data generated on a
big-endian machine. Add support for reading big-endian data arrays
from these files.

Commit:
Fixing Windows compile error

Addresses paraview/paraview#17767

mathstuf committed:
avtMFIXCDFFileFormat: remove netcdfcpp dependency

The netcdfcpp project is deprecated. Rather than porting to netcdf-c++,
just use the C API.

mathstuf committed:
visit_gzstream: fix build without libz support

mathstuf committed:
BOV: remove executable permissions

biagas committed:
Merge branch 'develop' into task/biagas/pr3456_update_with_develop

biagas committed:
Merge branch 'develop' into task/biagas/pr3456_update_with_develop

biagas committed:
Fix segv accesing beyond end of char array.

biagas commented:
@markcmiller86 Would you please check the H5F_CLOSE_SEMI additions in 
avtChomboFileFormat.C, avtEnzoFileFormat.C.

markcmiller86 commented:
> @markcmiller86 Would you please check the H5F_CLOSE_SEMI additions in
> avtChomboFileFormat.C, avtEnzoFileFormat.C.

Ok, just did. Looks fine to me. Both files `#include <visit-hdf5.h>` which uses CPP macros to map `H5Fopen()` and `H5Fclose()` calls to `VisIt_H5Fopen()` and `VisIt_H5Fclose()` call functions defined there. They could have opted NOT to define the `vtkFileCloseSemiProperty` class and simply used `H5P_DEFAULT` as the argument wherever it is used instead and achieved same behavior. It some ways, it is a bit odd that they go to the trouble of handling the close-semi property there but then don't also go to the trouble of checking the return value of `H5Fclose()` at their level too. But, since instances of that function are mapped to `VisIt_H5Fclose()` *and* that function does that work (plus some other useful debugging stuff), I think it is ok.

mathstuf commented:
> but then don't also go to the trouble of checking the return value of H5Fclose() at their level too.

It's a destructor. There's not much *to* do AFAIK. Though I'm not that familiar with HDF5. @danlipsa Do you have any input on this bit?

markcmiller86 commented:
I am pretty sure the only reason for setting close-semi is so that HDF5 will report errors on `H5Fclose()`. Hence, IMHO, setting close-semi but not then also checking return value of `H5Fclose()`everywhere it is used is potentially confusing to the reader of this code.

That said, it doesn't really matter as far as *functionality* in this case because by virtue of `#include visit-hdf5.h`, the CPP is mapping all those calls to `VisIt_H5Fclose()` which does check return value and check for left open objects. It just isn't obvious.

If there is any action to be taken here (and I am not suggesting there is), I would probably replace `vtkFileCloseSemiProperty` with `H5P_DEFAULT` everywhere and remove `vtkFileCloseSemiProperty` and just let the macros in `visit-hdf5.h` handle it. That is what we do in most other HDF5 plugins.

biagas commented:
I like the idea of consistent handling of HDF5 things within VisIt. @markcmiller86, what is the difference between H5P_DEFAULT (the previous use) and HDF5_DEFAULT?
 @mathstuf, do you have any objections to Mark's suggestion?  I will make the change if not.

biagas committed:
MFIXCDF is no longer dependent on NECTCDF_CXX lib.

Update xml, CMakeLists.txt and databases/CMakeList.txt to reflect this.

markcmiller86 commented:
> VisIt. @markcmiller86, what is the difference between H5P_DEFAULT (the previous use) and HDF5_DEFAULT

Sorry...that was a typo...in all cases, I mean `H5P_DEFAULT`.

biagas commented:
Given that the macros in visit-hdf5.h utilize H5F_CLOSE_SEMI, I'm wondering why there was the need to make the changes in avtChomboFileFormat.C and avtEnzoFileFormat.C, @mathstuf can you speak to that? The macros are codeblocked by these preprocesor defines:
```
#ifndef VISIT_DONT_REDEFINE_H5FOPENCLOSE
#ifdef H5_VERS_MAJOR
```
Is it that paraview's use-case isn't utilizing the macros due to these defines?

mathstuf commented:
I don't know the motivation behind the original change. I suspect some warning was getting dumped out somewhere, someone complained, and this was the solution? I don't see any issues about it. Searching my email archives also comes up empty, but I may not have been on the relevant thread (presumably with some customer).

biagas committed:
Merge branch 'develop' into task/biagas/pr3456_update_with_develop

biagas committed:
Revert changes.

After comments from @markcmiller86 , reverting the changes adding vtkFileCloseSemiProperty because the same functionality it added is contained in the visit-hdf5.h macros which redefine H5Fopen and H5Fclose.

biagas commented on src/avt/Database/Database/avtGenericDatabase.C:
```diff
@@ -3172,21 +3172,6 @@ avtGenericDatabase::GetMesh(const char *meshname, int ts, int domain,
-        }
-        tp->SetOutput(mesh);
-        tp->SetOutput(NULL);
-
```
@brugger1 I'm still not sure if this section of code is still needed in VisIt. I tried comparing timings logs from the test suite, but that wasn't very productive. Is there any other empirical study of performance that can be done on VisIt with this code intact and without it?
Or should I put the code back in, but comment out with a caveat it can be uncommented if performance seems poor?

mathstuf replied:
The commit message for this mentions that the new pipeline API (probably 8.x based on the 2017 authorship date) doesn't work with this pattern. @berkgeveci or @cjh1?

mathstuf replied:
Ah, we discussed this in #3456 already. I never got around to testing with this put back. I just rebased our branch on `develop` as it is today. Once that is in, I can test putting this back.

brugger1 commented:
It seems like you’d have to catch it when it does the garbage collection for a large number of data sets to assess the impact, which seems hard. I’d leave in the commented code and create a ticket to remove it on a certain date (a year or 2 from now) if there are no issues.

biagas committed:
Reinstate trival producer logic, but #if 0' out.

Will enter a ticket to fully remove at a later date if no performance penalty is observed.

biagas committed:
Merge branch 'develop' into task/biagas/pr3456_update_with_develop

biagas committed:
Merge branch 'develop' into task/biagas/pr3456_update_with_develop

biagas committed:
Merge branch 'develop' into task/biagas/pr3456_update_with_develop

biagas committed:
Merge branch 'develop' into task/biagas/pr3456_update_with_develop
